File: examples/c/deutsch-jozsa.c
#define _GNU_SOURCE
#define _USE_MATH_DEFINES
#include <qblaze.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#define ORACLE_N 4

static uint64_t rand64(void) {
	uint64_t val;
	int r = getentropy(&val, sizeof(val));
	if (r < 0) {
		perror("getentropy");
		abort();
	}
	return val;
}

int deutsch_jozsa(int (*oracle)(QBlazeSimulator*)) {
	QBlazeSimulator *sim = qblaze_new(NULL);
	if (!sim) return QBLAZE_ERR_MEMORY;

	int r = qblaze_apply_u3(sim, ORACLE_N, M_PI, M_PI, 0); // X
	if (r < 0) goto fail;

	for (size_t i = 0; i <= ORACLE_N; i++) {
		r = qblaze_apply_u3(sim, i, M_PI_2, 0, M_PI); // H
		if (r < 0) goto fail;
	}

	r = oracle(sim);
	if (r < 0) goto fail;

	for (size_t i = 0; i < ORACLE_N; i++) {
		r = qblaze_apply_u3(sim, i, M_PI_2, 0, M_PI); // H
		if (r < 0) goto fail;
	}

	double prob = 1.0;
	size_t num_ones = 0;

	for (size_t i = 0; i < ORACLE_N; i++) {
		double p0, p1;
		int r = qblaze_measure(sim, i, rand64(), &p0, &p1);
		if (r < 0) goto fail;
		if (r) {
			num_ones++;
			prob *= p1;
		} else {
			prob *= p0;
		}
	}

	qblaze_del(sim);

	const char *meaning = num_ones ? "balanced" : "constant";

	printf("Measured %zu ones (%s), likelihood %lf\n", num_ones, meaning, prob);
	return 0;

fail:
	qblaze_del(sim);
	return r;
}

static int constant_oracle(QBlazeSimulator *sim) {
	return qblaze_apply_u3(sim, ORACLE_N, M_PI, M_PI, 0); // X
}

static int balanced_oracle(QBlazeSimulator *sim) {
	int r;
	r = qblaze_apply_mcx(sim, (struct QBlazeControl[]){{0, true}, {1, false}}, 2, 2);
	if (r < 0) return r;
	r = qblaze_apply_mcx(sim, (struct QBlazeControl[]){{2, true}}, 1, ORACLE_N);
	if (r < 0) return r;
	r = qblaze_apply_mcx(sim, (struct QBlazeControl[]){{0, true}, {1, false}}, 2, 2);
	if (r < 0) return r;

	return qblaze_apply_mcx(sim, (struct QBlazeControl[]){{3, true}}, 1, ORACLE_N);
}

int main() {
	int r;

	printf("Constant oracle\n");
	r = deutsch_jozsa(constant_oracle);
	if (r < 0) goto fail;

	printf("Balanced oracle\n");
	r = deutsch_jozsa(balanced_oracle);
	if (r < 0) goto fail;

	return 0;
fail:
	fprintf(stderr, "Error %d\n", r);
	return -1;
}

File: examples/py/deutsch_jozsa.py
import typing

import qblaze


ORACLE_N = 4


def deutsch_jozsa(oracle: typing.Callable[[qblaze.Simulator], None]) -> None:
    sim = qblaze.Simulator()

    sim.x(ORACLE_N)

    for i in range(ORACLE_N + 1):
        sim.h(i)

    oracle(sim)

    for i in range(ORACLE_N):
        sim.h(i)

    prob = 1.0
    num_ones = 0
    for i in range(ORACLE_N):
        (r, p0, p1) = sim.measure_ext(i)
        if r:
            prob *= p1
            num_ones += 1
        else:
            prob *= p0

    meaning = f'balanced' if num_ones else 'constant'
    print(f'Measured {num_ones} ones ({meaning}), likelihood {prob:f}')


def constant_oracle(sim: qblaze.Simulator) -> None:
    sim.x(ORACLE_N)


def balanced_oracle(sim: qblaze.Simulator) -> None:
    sim.ccx(0, 1, 2)
    sim.cx(2, ORACLE_N)
    sim.ccx(0, 1, 2)
    sim.cx(3, ORACLE_N)


print('Constant oracle')
deutsch_jozsa(constant_oracle)
print('Balanced oracle')
deutsch_jozsa(balanced_oracle)
